src/guard.rs

```diff
@@ -0,0 +1,160 @@
+//!
+//! ```
+//! # use safe_transmute::{GuardError, ErrorReason};
+//! # use safe_transmute::guard::{Guard, PedanticGuard};
+//! assert_eq!(
+//!     PedanticGuard::check::<i16>(b"covfefe"),
+//!     Err(GuardError {
+//!         required: 2,
+//!         actual: 7,
+//!         reason: ErrorReason::InexactByteCount,
+//!     }));
+//! ```
+//!
+//! [`GuardError`]: ../type.GuardError.html
+//! [`PedanticGuard`]: struct.PedanticGuard.html
+//! [`PermissiveGuard`]: struct.PermissiveGuard.html
+
+use error::{ErrorReason, GuardError};
+use std::mem::align_of;
+
+/// The `Guard` type describes types which define boundary checking strategies.
+pub trait Guard {
+    /// Check the size of the given byte slice against a particular type.
+    ///
+    /// # Errors
+    ///
+    /// If the slice's size does not comply with this guard, an error
+    /// which specifies the incompatibility is returned.
+    fn check<T>(v: &[u8]) -> Result<(), GuardError>;
+}
+
+/// Single value guard: The byte slice must have exactly enough bytes to fill a single
+/// instance of a type.
+pub struct SingleValueGuard;
+impl Guard for SingleValueGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() != align_of::<T>() {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Pedantic guard: The byte slice must have at least enough bytes to fill a single
+/// instance of a type, and should not have extraneous data.
+pub struct PedanticGuard;
+impl Guard for PedanticGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() < align_of::<T>() {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::NotEnoughBytes,
+            })
+        } else if bytes.len() % align_of::<T>() != 0 {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A strict guard: The byte slice should not have extraneous data, but can be
+/// empty.
+pub struct StrictGuard;
+impl Guard for StrictGuard {
+    fn check<T>(bytes: &[u8]) -> Result<(), GuardError> {
+        if bytes.len() % align_of::<T>() != 0 {
+            Err(GuardError {
+                required: align_of::<T>(),
+                actual: bytes.len(),
+                reason: ErrorReason::InexactByteCount,
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A basic reasonable guard: The byte slice must have at least enough bytes to fill a single
+/// instance of a type, and extraneous data is ignored.
+pub struct BasicGuard;
```

Comment:
Not sure about "basic" as name -- it's semantically pretty void (i.e. it doesn't mean "will acquiesce at least 1 instance").